relay-server/src/metrics_extraction/metrics_summary.rs

```diff
@@ -0,0 +1,446 @@
+        let mut aggregator = MetricsSummaryAggregator::new();
+
+        for bucket in buckets {
+            aggregator.add(bucket);
+        }
+
+        aggregator
+    }
+
+    /// Adds a [`Bucket`] into the [`MetricsSummaryAggregator`].
+    pub fn add(&mut self, bucket: &Bucket) {
+        let key = MetricsSummaryBucketKey {
+            metric_name: bucket.name.clone(),
+            tags: bucket.tags.clone(),
+        };
+
+        let value = MetricsSummaryBucketValue::from_bucket_value(&bucket.value);
+
+        match self.buckets.entry(key) {
+            Entry::Occupied(mut entry) => {
+                entry.get_mut().merge(value);
+            }
+            Entry::Vacant(entry) => {
+                entry.insert(value);
+            }
+        }
+    }
+
+    /// Builds the [`MetricsSummary`] from the [`MetricsSummaryAggregator`].
+    ///
+    /// Note that this method consumes the aggregator itself, since the purpose of the aggregator
+    /// is to be built and destroyed once the summary is ready to be computed.
+    pub fn build_metrics_summary(self) -> MetricsSummary {
+        let mut metrics_summary = BTreeMap::new();
+
+        for (key, value) in self.buckets {
+            let tags = key
+                .tags
+                .into_iter()
+                .map(|(tag_key, tag_value)| (tag_key, Annotated::new(tag_value)))
+                .collect();
+
+            let metric_summary = MetricSummary {
+                min: value
+                    .min
+                    .map_or(Annotated::empty(), |m| Annotated::new(m.to_f64())),
+                max: value
+                    .max
+                    .map_or(Annotated::empty(), |m| Annotated::new(m.to_f64())),
+                sum: value
+                    .sum
+                    .map_or(Annotated::empty(), |m| Annotated::new(m.to_f64())),
+                count: Annotated::new(value.count),
+                tags: Annotated::new(tags),
+            };
+
+            let entry = metrics_summary
+                .entry(key.metric_name.to_string())
+                .or_insert(Annotated::new(vec![]));
+
+            if let Some(x) = entry.value_mut() {
+                x.push(Annotated::new(metric_summary));
+            }
+        }
+
+        MetricsSummary(metrics_summary)
+    }
+
+    /// Returns `true` if the [`MetricsSummaryAggregator`] is empty, `false` otherwise.
+    fn is_empty(&self) -> bool {
+        self.buckets.is_empty()
+    }
+}
+
+/// Computes the [`MetricsSummary`] from a slice of [`Bucket`]s.
+pub fn compute(buckets: &[Bucket]) -> Option<MetricsSummary> {
+    if buckets.is_empty() {
+        return None;
+    }
+
+    // For now, we only want metrics summaries to be extracted for custom metrics.
+    let filtered_buckets = buckets
+        .iter()
+        .filter(|b| matches!(b.name.namespace(), MetricNamespace::Custom));
+
+    let aggregator = MetricsSummaryAggregator::from_buckets(filtered_buckets);
+    if aggregator.is_empty() {
```

Comment:
We do the check in the aggregator directly so that we avoid collecting the filtered buckets into a vector just for the sake of determining the size.